File: sources/webcore/opt/webcore/libexec/account.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# --------------------------------
# Copyright (c) 2011 "Capensis" [http://www.capensis.com]
#
# This file is part of Canopsis.
#
# Canopsis is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Canopsis is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Canopsis.  If not, see <http://www.gnu.org/licenses/>.
# ---------------------------------

import sys
import os
import clogging
import json
import bottle

from bottle import route, get, put, delete, post
from bottle import request, HTTPError, response, redirect

# Canopsis
from caccount import caccount
from cstorage import cstorage
from cstorage import get_storage
from crecord import crecord

try:
	from cgroup import cgroup
except:
	pass

#import protection function
from libexec.auth import get_account, delete_session, reload_account, check_group_rights


logger = clogging.getLogger()

#group who have right to access 
group_managing_access = ['group.CPS_account_admin']

root_account = caccount(user="root", group="root")
#########################################################################

#### GET Me
@get('/account/me')
def account_get_me():
	ctype= 'account'
	
	#get the session (security)
	account = get_account()

	storage = get_storage(namespace='object')

	#try:
	logger.debug(" + Try to get '%s' ... " % account._id)

	try:
		record = storage.get(account._id, account=account)
	except:
		return HTTPError(404, 'Impossible to find account')

	#logger.debug("   + Result: '%s'" % record)
	#except Exception, err:
	#	self.logger.error("Exception !\nReason: %s" % err)
	#	return HTTPError(404, _id+" Not Found")

	if record:
		data = record.dump(json=True)
		data['id'] = data['_id']
		output = [data]
		reload_account(account._id)

	output={'total': 1, 'success': True, 'data': output}

	logger.debug(" + Output: " + str(output))
	
	logger.debug('Response status: %s' % response.status)
	
	return output

#### Get Avatar
@get('/account/getAvatar/:_id')
@get('/account/getAvatar')
def account_get_avatar(_id=None):
	account = get_account()
	storage = get_storage(namespace='object', account=account)

	if not _id:
		_id = account._id

	logger.debug('getAvatar of: %s' % _id)

	record = storage.get(_id, account=caccount(user="root", group="root"))
	
	if not record or not record.data.get('avatar_id', None):
		redirect("/static/canopsis/widgets/stream/logo/ui.png")

	avatar_id =  record.data.get('avatar_id')

	logger.debug(' + avatar_id: %s' % avatar_id)

	redirect("/files/%s" % avatar_id)

#### POST setConfig
@post('/account/setConfig/:_id')
def account_setConfig(_id):
	account = get_account()
	storage = get_storage(namespace='object')
	
	value = request.params.get('value', default=None)

	logger.debug(" + setConfig '%s' => '%s'" % (_id, value))

	if value:
		if _id == "shadowpasswd":
			account.shadowpasswd = value
		else:
			account.data[_id] = value

		storage.put(account, account=account)
		output={'total': 0, 'success': True, 'data': []}
	else:
		output={'total': 0, 'success': False, 'data': []}
	
	return output

@get('/account/getAuthKey/:dest_account')
def account_getAuthKey(dest_account):
	if not dest_account:
		return HTTPError(404, 'No account specified')
	
	#------------------------get accounts----------------------
	account = get_account()
	storage = get_storage(namespace='object',account=account)
	
	_id = 'account.%s' % dest_account
	
	try:
		aim_account = caccount(storage.get(_id,account=account))
		
		return {'total':1,'success':True,'data':{'authkey':aim_account.get_authkey()}}
	except Exception,err:
		logger.debug('Error while fetching account : %s' % err)
		return {'total':0,'success':False,'data':{'output':str(err)}}
		
	

@get('/account/getNewAuthKey/:dest_account', checkAuthPlugin={'authorized_grp':'group.CPS_authkey'})
def account_newAuthKey(dest_account):
	if not dest_account:
		return HTTPError(404, 'No account specified')
	
	#------------------------get accounts----------------------
	account = get_account()
	storage = get_storage(namespace='object',account=account)
	
	_id = 'account.%s' % dest_account
	
	try:
		aim_account = caccount(storage.get(_id,account=account))
	except:
		logger.debug('aimed account not found')
		return HTTPError(404, 'Wrong account name or no enough rights')
	
	#---------------------generate new key-------------------
	logger.debug('Change AuthKey for : %s' % aim_account.user)
	
	try:
		aim_account.generate_new_authkey()
		storage.put(aim_account,account=account)
		logger.debug('New auth key is : %s' % aim_account.get_authkey())
		return {'total': 0, 'success': True, 'data': {'authkey': aim_account.get_authkey(),'account':aim_account.user}}
	except Exception,err:
		logger.error('Error while updating auth key : %s' % err)
		return {'total': 0, 'success': False, 'data': {}}
	
#### GET
@get('/account/:_id')
@get('/account/')
def account_get(_id=None):
	namespace = 'object'
	ctype= 'account'
	
	#get the session (security)
	account = get_account()

	if not check_group_rights(account, 'group.account_managing'):
		return HTTPError(403, 'Insufficient rights')

	limit = int(request.params.get('limit', default=20))
	#page =  int(request.params.get('page', default=0))
	start =  int(request.params.get('start', default=0))
	#groups = request.params.get('groups', default=None)
	search = request.params.get('search', default=None)

	logger.debug("GET:")
	logger.debug(" + User: "+str(account.user))
	logger.debug(" + Group(s): "+str(account.groups))
	logger.debug(" + namespace: "+str(namespace))
	logger.debug(" + Ctype: "+str(ctype))
	logger.debug(" + _id: "+str(_id))
	logger.debug(" + Limit: "+str(limit))
	#logger.debug(" + Page: "+str(page))
	logger.debug(" + Start: "+str(start))
	#logger.debug(" + Groups: "+str(groups))
	logger.debug(" + Search: "+str(search))

	storage = get_storage(namespace=namespace)

	total = 0
	mfilter = {}
	if ctype:
		mfilter = {'crecord_type': ctype}
	if _id:	
		try:
			records = [ storage.get(_id, account=account) ]
			total = 1
		except Exception, err:
			self.logger.error("Exception !\nReason: %s" % err)
			return HTTPError(404, _id+" Not Found")
		
	else:
		records =  storage.find(mfilter, limit=limit, offset=start, account=account)
		total =	   storage.count(mfilter, account=account)

	output = []
	for record in records:
		if record:
			data = record.dump(json=True)
			data['id'] = data['_id']
			output.append(data)

	output={'total': total, 'success': True, 'data': output}

	#logger.debug(" + Output: "+str(output))

	return output

	
#### POST
@post('/account/', checkAuthPlugin={'authorized_grp':group_managing_access})
def account_post():
	#get the session (security)
	account = get_account()
	storage = get_storage(namespace='object',account=account)

	logger.debug("POST:")

	data = request.body.readline()
	if not data:
		return HTTPError(400, "No data received")

	data = json.loads(data)

	## Clean data
	try:
		del data['_id']
		del data['id']
		del data['crecord_type']
	except:
		pass
	
	if data['user']:
		#check if already exist
		already_exist = False
		_id = "account.%s" % data['user']
		try:
			record = storage.get(_id, account=account)
			logger.debug('Update account %s' % _id)
			already_exist = True
		except:
			logger.debug('Create account %s' % _id)
			
		if already_exist:
			return HTTPError(405, "Account already exist, use put method for update !")

		#----------------------------CREATION--------------------------
		create_account(data)

	else:
		logger.warning('WARNING : no user specified ...')
		
@put('/account/',checkAuthPlugin={'authorized_grp':group_managing_access})
@put('/account/:_id',checkAuthPlugin={'authorized_grp':group_managing_access})
def account_update(_id=None):
	account = get_account()
	storage = get_storage(namespace='object', account=account)
	
	logger.debug("PUT:")

	data = request.body.readline()
	if not data:
		return HTTPError(400, "No data received")
	data = json.loads(data)
	
	if not isinstance(data,list):
		data = [data]

	for item in data:
		logger.debug(item)
		if '_id' in item:
			_id = item['_id']
			del item['_id']
		if 'id' in item:
			_id = item['id']
			del item['id']

		if not _id:
			return HTTPError(400, "No id recieved")
		
		try:
			record = caccount(storage.get(_id ,account=account))
			logger.debug('Update account %s' % _id)
		except:
			logger.debug('Account %s not found' % _id)
			return HTTPError(404, "Account to update not found")
		
		#Get password
		if 'passwd' in item:
			logger.debug(' + Update password ...')
			record.passwd(str(item['passwd']))
			del item['passwd']
			
		#Get group
		if 'aaa_group' in item:
			logger.debug(' + Update group ...')
			record.chgrp(str(item['aaa_group']))
			del item['aaa_group']
				
		#get secondary groups
		if 'groups' in item:
			groups = []
			for group in item['groups']:
				if group.find('group.') == -1:
					groups.append('group.%s' % group)
				else:
					groups.append(group)

			logger.debug(' + Update groups ...')	
			logger.debug(' + Old groups : %s' % str(record.groups))
			logger.debug(' + New groups : %s' % str(groups))
			record.groups = groups
			del item['groups']
		

		for _key in item:
			logger.debug('Update %s with %s' % (str(_key),item[_key]))
			setattr(record,_key,item[_key])

		storage.put(record, account=account)
		reload_account(record._id)
	
#### DELETE
@delete('/account/',checkAuthPlugin={'authorized_grp':group_managing_access})
@delete('/account/:_id',checkAuthPlugin={'authorized_grp':group_managing_access})
def account_delete(_id=None):
	account = get_account()
	storage = get_storage(namespace='object')
	
	logger.debug("DELETE:")

	data = request.body.readline()
	if data:
		try:
			data = json.loads(data)
		except:
			logger.warning('Invalid data in request payload')
			data = None

	if data:
		logger.debug(" + Data: %s" % data)

		if isinstance(data, list):
			logger.debug(" + Attempt to remove %i item from db" % len(data))
			_id = []
				
			for item in data:
				if isinstance(item,str):
					_id.append(item)
					
				if isinstance(item,dict):
					item_id = item.get('_id', item.get('id', None))
					if item_id:
						_id.append(item_id)

		if isinstance(data, str):
			_id = data

		if isinstance(data, dict):
			_id = data.get('_id', data.get('id', None))

	if not _id:
		return HTTPError(404, "No '_id' field in header ...")
	
	logger.debug(" + _id: %s " % _id)
	try:
		storage.remove(_id, account=account)
		delete_session(_id)
	except:
		return HTTPError(404, _id+" Not Found")

	#delete all object
	if not isinstance(_id, list):
		_id = [_id]

	mfilter = {'aaa_owner':{'$in':_id}}
	record_list = storage.find(mfilter=mfilter,account=account)
	record_id_list = [record._id for record in record_list]

	try:
		storage.remove(record_id_list,account=account)
	except Exception as err:
		log.error('Error While suppressing account items: %s' % err)

	logger.debug('account removed')

### GROUP
@post('/account/addToGroup/:group_id/:account_id',checkAuthPlugin={'authorized_grp':group_managing_access})
def add_account_to_group(group_id=None,account_id=None):
	session_account = get_account()
	storage = get_storage(namespace='object', account=session_account)
	
	if not group_id or not account_id:
		return HTTPError(400, 'Bad request, must specified group and account')
	
	#get group && account
	if group_id.find('group.') == -1:
		group_id = 'group.%s' % group_id
		
	if account_id.find('account.') == -1:
		account_id = 'account.%s' % account_id
		
	logger.debug('Try to get %s and %s' % (account_id,group_id))
		
	try:
		account_record = storage.get(account_id, account=session_account)
		account = caccount(account_record)
		group_record = storage.get(group_id, account=session_account)
		group = cgroup(group_record)
		
	except Exception,err:
		logger.error('error while fetching %s and %s : %s' % (account_id,group_id,err))
		return HTTPError(403, 'Record not found or insufficient rights')
		
	#put in group
	group.add_accounts(account)
	
	try:
		storage.put([group,account])
	except:
		logger.error('Put group/account in db goes wrong')
		return HTTPError(500, 'Put group/account in db goes wrong')
	
	return {'total' :1, 'success' : True, 'data':[]}
		
@post('/account/removeFromGroup/:group_id/:account_id',checkAuthPlugin={'authorized_grp':group_managing_access})
def remove_account_from_group(group_id=None,account_id=None):
	session_account = get_account()
	storage = get_storage(namespace='object',account=session_account)
	
	if not group_id or not account_id:
		return HTTPError(400, 'Bad request, must specified group and account')
	
	#get group && account
	if group_id.find('group.') == -1:
		group_id = 'group.%s' % group_id
		
	if account_id.find('account.') == -1:
		account_id = 'account.%s' % account_id
		
	logger.debug('Try to get %s and %s' % (account_id,group_id))
		
	try:
		account_record = storage.get(account_id,account=session_account)
		account = caccount(account_record)
		group_record = storage.get(group_id,account=session_account)
		group = cgroup(group_record)
		
	except Exception,err:
		logger.error('error while fetching %s and %s : %s' % (account_id,group_id,err))
		return HTTPError(403, 'Record not found or insufficient rights')
		
	#remove in group
	
	group.remove_accounts(account)
	
	try:
		storage.put([group,account])
	except:
		logger.error('Put group/account in db goes wrong')
		return HTTPError(500, 'Put group/account in db goes wrong')
	
	return {'total' :1, 'success' : True, 'data':[]}


def create_account(data):
	logger.debug(' + New account')
	new_account = caccount(
		user=data['user'],
		group=data.get('aaa_group', None),
		lastname=data['lastname'],
		firstname=data['firstname'],
		mail=data['mail']
	)

	new_account.external = data.get('external', False)

	#passwd
	passwd = data['passwd']
	new_account.passwd(passwd)
	logger.debug("   + Passwd: '%s'" % passwd)

	#secondary groups
	groups = []
	for group in data.get('groups', []):
		if group.find('group.') == -1:
			groups.append('group.%s' % group)
		else:
			groups.append(group)
	new_account.groups = groups
	
	storage = get_storage(namespace='object')

	#put record
	logger.debug(' + Save new account')
	new_account.chown(new_account._id)
	storage.put(new_account, account=root_account)
	
	#get rootdir
	logger.debug(' + Create view directory')
	rootdir = storage.get('directory.root', account=root_account)
	
	if rootdir:
		userdir = crecord({'_id': 'directory.root.%s' % new_account.user,'id': 'directory.root.%s' % new_account.user ,'expanded':'true'}, type='view_directory', name=new_account.user)
		userdir.chown(new_account._id)
		userdir.chgrp(new_account.group)
		userdir.chmod('g-w')
		userdir.chmod('g-r')

		rootdir.add_children(userdir)
		storage.put([rootdir,userdir], account=root_account)
	else:
		logger.error('Impossible to get rootdir')

	return new_account

File: community/sources/webcore/var/www/canopsis/app/controller/Account.js
//need:app/lib/controller/cgrid.js,app/view/Account/Grid.js,app/view/Account/Form.js,app/store/Accounts.js
/*
# Copyright (c) 2011 "Capensis" [http://www.capensis.com]
#
# This file is part of Canopsis.
#
# Canopsis is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Canopsis is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with Canopsis.  If not, see <http://www.gnu.org/licenses/>.
*/
Ext.define('canopsis.controller.Account', {
	extend: 'canopsis.lib.controller.cgrid',

	views: ['Account.Grid', 'Account.Form'],
	stores: ['Accounts'],
	models: ['Account'],

	iconCls: 'icon-crecord_type-account',

	logAuthor: '[controller][Account]',

	init: function() {
		log.debug('[' + this.id + '] - Initialize ...');

		this.formXtype = 'AccountForm';
		this.listXtype = 'AccountGrid';

		this.modelId = 'Account';

		this.callParent(arguments);

		global.accountCtrl = this;
	},

	bindGridEvents: function(grid){
		var ldapCtrl = this.getController('Ldap');

		// Bind ldap button
		var btns = Ext.ComponentQuery.query('#' + grid.id + ' button[action=ldap]');

		for(var i = 0; i < btns.length; i++) {
			btns[i].on('click', ldapCtrl.accountButton, ldapCtrl);
		}
	},

	getConfig: function(id, default_value) {
		log.debug(' + getConfig ' + id, this.logAuthor);
		if(!global.account[id]) {
			if(global[id]) {
				global.account[id] = global[id];
			}
			else {
				global.account[id] = default_value;
			}
		}

		return global.account[id];
	},

	setConfig: function(id, value, cb) {
		log.debug(' + setConfig ' + id + ' => ' + value, this.logAuthor);
		global.account[id] = value;

		var url = '/account/setConfig/' + id;

		if(!cb) {
			cb = function() {
				log.debug(' + setConfig Ok', this.logAuthor);
			};
		}

		ajaxAction(url, {value: value}, cb, this, 'POST');

		return global.account[id];
	},

	_deleteButton: function(button) {
		log.debug('Clicked deleteButton', this.logAuthor);
		var grid = this.grid;
		var me = this;

		var selection = grid.getSelectionModel().getSelection();

		if(selection) {
			//check right
			var ctrlAccount = this.getController('Account');
			var authorized = true;

			for(var i = 0; i < selection.length; i++) {
				if(!ctrlAccount.check_record_right(selection[i], 'w')) {
					authorized = false;
				}

				if(this.checkInternal && selection[i].get('internal')) {
					authorized = false;
				}

				if(!authorized) {
					break;
				}
			}

			if(authorized === true) {
				Ext.Msg.show({
					title:_('Are you sure ?'),
					inputSelection: selection,
					msg: _('When you delete an account ALL its object (view, curve...) is deleted with it. You can choose to just disable the account in order to keep its objects.'),
					buttons: Ext.Msg.YESNOCANCEL,
					buttonText: {
						yes: _('DELETE'),
						no: _('Disable')
					},
					scope: me,
					fn: function(buttonId, text, opt) {
						void(text);

						if(buttonId === 'yes') {
							this.grid.store.remove(opt.inputSelection);
						}
						if(buttonId === 'no') {
							this._enabledisable();
						}
					}
				});
			}
			else {
				global.notify.notify(_('Access denied'), _('You don\'t have the rights to modify this object'), 'error');
			}
		}

		if(this.deleteButton) {
			this.deleteButton(button, grid, selection);
		}

	},

	beforeload_DuplicateForm: function(form, copy) {
		void(form);

		/* remove unique data */
		copy.set('user', undefined);
		copy.set('firstname', undefined);
		copy.set('lastname', undefined);
		copy.set('mail', undefined);
	},

	afterload_DuplicateForm: function(form, copy) {
		this.form_loadGroups(form, copy);
	},

	logout: function() {
		Ext.Ajax.request({
			url: '/logout',
			scope: this,
			success: function() {
				log.debug(' + Success.', this.logAuthor);
				window.location.reload();
			},
			failure: function() {
				log.error("Logout impossible, maybe you're already logout");
			}
		});
	},

	setLocale: function(locale) {
		var cb = function() {
			log.debug(' + setLocale Ok', this.logAuthor);

			Ext.MessageBox.show({
				title: _('Configure language'),
				msg: _('Application must be reloaded, do you want to reload now ?'),
				icon: Ext.MessageBox.WARNING,
				buttons: Ext.Msg.OKCANCEL,
				fn: function(btn) {
					if(btn === 'ok') {
						if(global.minimified) {
							window.location.href = '/' + locale + '/';
						}
						else {
							window.location.href = '/' + locale + '/static/canopsis/index.debug.html';
						}
					}
				}
			});
		};

		Ext.util.Cookies.set('locale', locale);

		this.setConfig('locale', locale, cb);
	},

	setClock: function(clock_type) {
		var cb = function() {
			log.debug(' + setClock Ok', this.logAuthor);

			Ext.MessageBox.show({
				title: _('Configure clock display'),
				msg: _('Application must be reloaded, do you want to reload now ?'),
				icon: Ext.MessageBox.WARNING,
				buttons: Ext.Msg.OKCANCEL,

				fn: function(btn) {
					if(btn === 'ok') {
						window.location.reload();
					}
				}
			});
		};

		this.setConfig('clock_type', clock_type, cb);
	},

	beforeload_EditForm: function(form,item) {
		var pass_textfield = Ext.ComponentQuery.query('#' + form.id + ' textfield[name=passwd]')[0];

		if(pass_textfield) {
			pass_textfield.allowBlank = true;
		}

		var user_textfield = Ext.ComponentQuery.query('#' + form.id + ' textfield[name=user]')[0];

		if(user_textfield) {
			user_textfield.hide();
		}

		// groups
		this.form_loadGroups(form, item);
	},

	form_loadGroups: function(form, item) {
		var store = Ext.getStore('Groups');
		var groups = item.get('groups');

		if(groups.length > 0) {
			var groups_records = [];

			for(var i = 0; i < groups.length; i++) {
				record = store.findExact('_id', groups[i]);

				if(record !== -1) {
					groups_records.push(store.getAt(record));
				}
			}

			var groups_grid = form.down('grid');
			groups_grid.getSelectionModel().select(groups_records, false, true);
		}
	},

	preSave: function(record, data, form) {
		//don't update password if it's empty
		if(form.editing && (record.get('passwd') === '')) {
			delete record.data.passwd;
		}

		if(record.get('external') === true) {
			log.debug('Impossible to change password on external account', this.logAuthor);
			delete record.data.passwd;
		}

		//add groups
		var checkGrid = form.down('grid');
		var record_list = checkGrid.getSelectionModel().getSelection();
		var groups = [];

		for(var i = 0; i < record_list.length; i++) {
			groups.push(record_list[i].get('crecord_name'));
		}

		record.set('groups', groups);

		//hack, otherwise webservice go wild (temporary)
		record.modified.aaa_group = data['aaa_group'];

		return record;
	},

	validateForm: function(store, data, form) {
		var already_exist = false;

		if ( !store ) { 
			var store = Ext.getStore('Groups');
		}

		// in creation mode
		if(!form.editing && store.findExact('user', data['user']) >= 0) {
			already_exist = true;
		}

		var field = form.findField('user');

		if(field) {
			field.markInvalid(_('Invalid field'));
		}

		if(already_exist) {
			log.debug('User already exist', this.logAuthor + '[validateForm]');
			global.notify.notify(data['user'] + ' already exist', 'you can\'t add the same user twice', 'error');

			return false;
		}
		else {
			return true;
		}
	},

	//check if user have right on this record
	check_record_right: function(record, option) {
		var user = global.account._id;
		var group = global.account.aaa_group;
		var groups = global.account.groups;

		//root can do everything
		if(this.checkRoot()) {
			return true;
		}

		if((option === 'r') || (option === 'w')) {
			if((user === record.get('aaa_owner')) && (Ext.Array.contains(record.data.aaa_access_owner, option))) {
				return true;
			}
			else if((group === record.get('aaa_group')) && (Ext.Array.contains(record.data.aaa_access_group, option))) {
				return true;
			}
			else if((Ext.Array.contains(groups, record.get('aaa_group'))) && (Ext.Array.contains(record.data.aaa_access_group, option))) {
				return true;
			}
			else if((Ext.Array.contains(groups, record.get('aaa_admin_group'))) || (group === record.get('aaa_admin_group'))) {
				return true;
			}
			else {
				return false;
			}
		}
		else {
			log.error(_('Incorrect right option given'), this.logAuthor);
		}
	},

	//check if user have right on this obj
	check_right: function(obj, option) {
		var user = global.account._id;
		var group = global.account.aaa_group;
		var groups = global.account.groups;

		//root can do everything
		if(this.checkRoot()) {
			return true;
		}

		if((option === 'r') || (option === 'w')) {
			if((user === obj.aaa_owner) && (Ext.Array.contains(obj.aaa_access_owner, option))) {
				return true;
			}
			else if((group === obj.aaa_group) && (Ext.Array.contains(obj.aaa_access_group, option))) {
				return true;
			}
			else if((Ext.Array.contains(groups, obj.aaa_group)) && (Ext.Array.contains(obj.aaa_access_group, option))) {
				return true;
			}
			else if((Ext.Array.contains(groups, obj.aaa_admin_group)) || group === obj.aaa_admin_group) {
				return true;
			}
			else {
				return false;
			}
		}
		else {
			log.error(_('Incorrect right option given'), this.logAuthor);
		}

	},

	// if the callback function is not null and ajax request is successful,
	// then the callback is called in passed scope with new key as argument.
	new_authkey: function(account, callback_func, scope) {
		if(account) {
			// ajax request
			log.debug('Ask webserver for new authentification key', this.logAuthor);

			Ext.Ajax.request({
				url: '/account/getNewAuthKey/' + account,
				method: 'GET',
				scope: scope,
				success: function(response) {
					var object_response = Ext.decode(response.responseText);

					if(object_response.success === true) {
						global.notify.notify(_('Success'), _('Your authentification key is updated'), 'success');

						var authkey = object_response.data.authkey;
						global.account.authkey = authkey;

						if(callback_func) {
							callback_func.call(this, authkey);
						}
					}
					else {
						log.error('Ajax output incorrect', this.logAuthor);
					}
				},
				failure: function() {
					global.notify.notify(_('Error'), _('An error have occured during the updating process'), 'error');
					log.error('Error while fetching new Authkey', this.logAuthor);
				}
			});
		}
		else {
			log.error('No account provided for Authkey');
		}
	},

	get_authkey: function(account, callback_func, scope) {
		log.debug('Ask webserver for authentification key', this.logAuthor);

		Ext.Ajax.request({
			url: '/account/getAuthKey/' + account,
			method: 'GET',
			scope: scope,
			success: function(response) {
				var object_response = Ext.decode(response.responseText);

				if(object_response.success === true) {
					var authkey = object_response.data.authkey;

					if(callback_func) {
						callback_func.call(this, authkey);
					}
				}
				else {
					log.error('Ajax output incorrect', this.logAuthor);
				}
			},
			failure: function() {
				global.notify.notify(_('Error'), _('An error have occured during the process'), 'error');
				log.error('Error while fetching new Authkey', this.logAuthor);
			}
		});
	},

	add_to_group: function(group, account) {
		log.debug('Ask webserver adding ' + account + ' to ' + 'group', this.logAuthor);

		if(group.search('group.') === -1) {
			group = 'group.' + group;
		}

		if (account.search('account.') === -1) {
			account = 'account.' + account;
		}

		Ext.Ajax.request({
			url: '/account/addToGroup/' + group + '/' + account,
			method: 'POST',
			success: function(response) {
				var object_response = Ext.decode(response.responseText);

				if(object_response.success === true) {
					global.notify.notify(_('Group added'), _('Group successfuly added to secondary groups'), 'success');
				}
				else {
					log.error('Ajax output incorrect', this.logAuthor);
				}
			},
			failure: function() {
				global.notify.notify(_('Error'), _('An error have occured during the process'), 'error');
				log.error('Error while fetching new Authkey', this.logAuthor);
			}
		});
	},

	remove_from_group: function(group, account) {
		log.debug('Ask webserver removing ' + account + ' from ' + 'group', this.logAuthor);

		Ext.Ajax.request({
			url: '/account/removeFromGroup/' + group + '/' + account,
			method: 'POST',
			success: function(response) {
				var object_response = Ext.decode(response.responseText);

				if(object_response.success === true) {
					global.notify.notify(_('Group removed'), _('Group successfuly removed from secondary groups'), 'success');
				}
				else {
					log.error('Ajax output incorrect', this.logAuthor);
				}
			},
			failure: function() {
				global.notify.notify(_('Error'), _('An error have occured during the process'), 'error');
				log.error('Error while fetching new Authkey', this.logAuthor);
			}
		});
	},

	checkGroup: function(group) {
		if(global.account.aaa_group === group || (Ext.Array.contains(global.account.groups, group))) {
			return true;
		}
		else {
			return false;
		}
	},

	checkRoot: function() {
		if(global.account.user === 'root' || global.account.aaa_group === 'group.CPS_root' || (Ext.Array.contains(global.account.groups, 'group.CPS_root'))) {
			return true;
		}
		else {
			return false;
		}
	},

	setPassword: function(password){
		var passwd = $.encoding.digests.hexSha1Str(password);

		this.setConfig("shadowpasswd", passwd, function() {
			global.notify.notify(_('Success'), _('Your password is updated'), 'success');
		});
	},

	setAvatar: function(file_id, filename) {
		var me = this;

		var extension = filename.split('.').pop();

		if($.inArray(extension, ['png', 'jpeg', 'jpg', 'gif']) === -1) {
			global.notify.notify(_('Failed'), "File extension not valid", 'error');
			return;
		}

		var avatar_id = file_id;

		log.debug('Set avatar_id in backend', this.logAuthor);

		Ext.Ajax.request({
			method: 'POST',
			url: '/account/setConfig/avatar_id',
			params: {
				value: avatar_id
			},
			success: function() {
				global.notify.notify(_('Success'), _('Avatar setted'), 'success');
				log.debug(' + Done', me.logAuthor);

				global.account.avatar_id = avatar_id;

				// Update icon in main bar
				log.debug('Refresh Mainbar', me.logAuthor);
				Ext.ComponentQuery.query('button[iconCls="icon-mainbar icon-avatar-bar"]')[0].setIcon('/account/getAvatar');
			},
			failure: function() {
				global.notify.notify(_('Failed'), _('Impossible to set this file as avatar'), 'error');
				log.error('Impossible to set this file as avatar', me.logAuthor);
			}
		});
	}
});
